### ISSUE TITLE
Null path exception in save process

### ISSUE THREAD

**Issue by shiftyscales** [Frooxius/NeosPublic#1308]
Reported by Medra, their Creator Jam world was unable to be saved properly. The dumped world can reproduce this issue reliably.
Log: [SHIFTYWINDOWS - 2020.11.8.605 - 2020-11-08 16_26_16.log](https://github.com/Frooxius/NeosPublic/files/5507452/SHIFTYWINDOWS.-.2020.11.8.605.-.2020-11-08.16_26_16.log)
The dumped world can be found here: https://ln2.sync.com/dl/f909edae0/jvk698xn-48xb57w5-9daf4ksw-zitb9fdd

To replicate:
- Import the qw3khrrcphkyyfvf.lz4bson file, and attempt to save the world

I will attempt to see if I can isolate a particular part of the world hierarchy responsible for the issue.

```
4:29:25 PM.152 (107 FPS)	Exception in the save process! Dumping the raw save data to: C:/Users/Shifty/AppData/Local/Temp/Solirax/NeosVR\Cache\9xcr0yzq315x7x4b.lz4bson
System.ArgumentNullException: Value cannot be null.
Parameter name: path
  at System.IO.FileStream..ctor (System.String path, System.IO.FileMode mode, System.IO.FileAccess access, System.IO.FileShare share, System.Int32 bufferSize, System.Boolean anonymous, System.IO.FileOptions options) [0x00014] in <9577ac7a62ef43179789031239ba8798>:0 
  at System.IO.FileStream..ctor (System.String path, System.IO.FileMode mode, System.IO.FileAccess access, System.IO.FileShare share) [0x00000] in <9577ac7a62ef43179789031239ba8798>:0 
  at (wrapper remoting-invoke-with-check) System.IO.FileStream..ctor(string,System.IO.FileMode,System.IO.FileAccess,System.IO.FileShare)
  at System.IO.File.OpenRead (System.String path) [0x00000] in <9577ac7a62ef43179789031239ba8798>:0 
  at BaseX.DataTreeConverter.Load (System.String file, System.String ext) [0x0009b] in <63fda06f32264f07a4cc10f7fa1ccde2>:0 
  at BaseX.DataTreeConverter.Load (System.String file, System.Uri uri) [0x0001b] in <63fda06f32264f07a4cc10f7fa1ccde2>:0 
  at FrooxEngine.DataTreeAssetProcessor+<Process>d__1.MoveNext () [0x001a2] in <70142345391c4379ae523c268651f353>:0 
--- End of stack trace from previous location where exception was thrown ---
  at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw () [0x0000c] in <9577ac7a62ef43179789031239ba8798>:0 
  at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess (System.Threading.Tasks.Task task) [0x0003e] in <9577ac7a62ef43179789031239ba8798>:0 
  at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification (System.Threading.Tasks.Task task) [0x00028] in <9577ac7a62ef43179789031239ba8798>:0 
  at System.Runtime.CompilerServices.TaskAwaiter.ValidateEnd (System.Threading.Tasks.Task task) [0x00008] in <9577ac7a62ef43179789031239ba8798>:0 
  at System.Runtime.CompilerServices.ConfiguredTaskAwaitable`1+ConfiguredTaskAwaiter[TResult].GetResult () [0x00000] in <9577ac7a62ef43179789031239ba8798>:0 
  at FrooxEngine.DataTreeSaver+<SaveLocally>d__11.MoveNext () [0x0007f] in <70142345391c4379ae523c268651f353>:0 
--- End of stack trace from previous location where exception was thrown ---
  at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw () [0x0000c] in <9577ac7a62ef43179789031239ba8798>:0 
  at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess (System.Threading.Tasks.Task task) [0x0003e] in <9577ac7a62ef43179789031239ba8798>:0 
  at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification (System.Threading.Tasks.Task task) [0x00028] in <9577ac7a62ef43179789031239ba8798>:0 
  at System.Runtime.CompilerServices.TaskAwaiter.ValidateEnd (System.Threading.Tasks.Task task) [0x00008] in <9577ac7a62ef43179789031239ba8798>:0 
  at System.Runtime.CompilerServices.ConfiguredTaskAwaitable`1+ConfiguredTaskAwaiter[TResult].GetResult () [0x00000] in <9577ac7a62ef43179789031239ba8798>:0 
  at FrooxEngine.Userspace+<SaveWorldTaskIntern>d__207.MoveNext () [0x002c1] in <70142345391c4379ae523c268651f353>:0 

   at void BaseX.UniLog.Error(string message, bool stackTrace)
   at async Task<Record> FrooxEngine.Userspace.SaveWorldTaskIntern(World world, Record record, RecordOwnerTransferer transferer)
   at void System.Runtime.CompilerServices.AsyncMethodBuilderCore+MoveNextRunner.InvokeMoveNext(object stateMachine)
   at void System.Runtime.CompilerServices.AsyncMethodBuilderCore+MoveNextRunner.Run()
   at void System.Threading.Tasks.AwaitTaskContinuation.RunOrScheduleAction(Action action, bool allowInlining, ref Task currentTask)
   at void System.Threading.Tasks.Task.FinishContinuations()
   at void System.Threading.Tasks.Task.FinishStageThree()
   at void System.Threading.Tasks.Task.FinishStageTwo()
   at void System.Threading.Tasks.Task.Finish(bool bUserDelegateExecuted)
   at bool System.Threading.Tasks.Task<Uri>.TrySetException(object exceptionObject)
   at void System.Runtime.CompilerServices.AsyncTaskMethodBuilder<Uri>.SetException(Exception exception)
   at async Task<Uri> FrooxEngine.DataTreeSaver.SaveLocally(SavedGraph graph, Uri uri)
   at void System.Runtime.CompilerServices.AsyncMethodBuilderCore+MoveNextRunner.InvokeMoveNext(object stateMachine)
   at void System.Runtime.CompilerServices.AsyncMethodBuilderCore+MoveNextRunner.Run()
   at void System.Threading.Tasks.AwaitTaskContinuation.RunOrScheduleAction(Action action, bool allowInlining, ref Task currentTask)
   at void System.Threading.Tasks.Task.FinishContinuations()
   at void System.Threading.Tasks.Task.FinishStageThree()
   at void System.Threading.Tasks.Task.FinishStageTwo()
   at void System.Threading.Tasks.Task.Finish(bool bUserDelegateExecuted)
   at bool System.Threading.Tasks.Task<Uri>.TrySetException(object exceptionObject)
   at void System.Runtime.CompilerServices.AsyncTaskMethodBuilder<Uri>.SetException(Exception exception)
   at async Task<Uri> FrooxEngine.DataTreeAssetProcessor.Process(SavedGraph graph, Uri graphUrl)
   at void System.Runtime.CompilerServices.AsyncMethodBuilderCore+MoveNextRunner.InvokeMoveNext(object stateMachine)
   at void System.Runtime.CompilerServices.AsyncMethodBuilderCore+MoveNextRunner.Run()
   at void System.Threading.Tasks.AwaitTaskContinuation.RunOrScheduleAction(Action action, bool allowInlining, ref Task currentTask)
   at void System.Threading.Tasks.Task.FinishContinuations()
   at void System.Threading.Tasks.Task.FinishStageThree()
   at bool System.Threading.Tasks.Task<string>.TrySetResult(string result)
   at void System.Runtime.CompilerServices.AsyncTaskMethodBuilder<string>.SetResult(string result)
   at async Task<string> FrooxEngine.DataTreeSaver.Gather(Uri uri)
   at void System.Runtime.CompilerServices.AsyncMethodBuilderCore+MoveNextRunner.InvokeMoveNext(object stateMachine)
   at void System.Runtime.CompilerServices.AsyncMethodBuilderCore+MoveNextRunner.Run()
   at void System.Threading.Tasks.AwaitTaskContinuation.RunOrScheduleAction(Action action, bool allowInlining, ref Task currentTask)
   at void System.Threading.Tasks.Task.FinishContinuations()
   at void System.Threading.Tasks.Task.FinishStageThree()
   at bool System.Threading.Tasks.Task<string>.TrySetResult(string result)
   at void System.Runtime.CompilerServices.AsyncTaskMethodBuilder<string>.SetResult(string result)
   at void System.Runtime.CompilerServices.AsyncValueTaskMethodBuilder<string>.SetResult(string result)
   at async ValueTask<string> FrooxEngine.AssetGatherer.Gather(Uri url, Priority priority)
   at void System.Runtime.CompilerServices.AsyncMethodBuilderCore+MoveNextRunner.InvokeMoveNext(object stateMachine)
   at void System.Runtime.CompilerServices.AsyncMethodBuilderCore+MoveNextRunner.Run()
   at void System.Threading.Tasks.AwaitTaskContinuation.System.Threading.IThreadPoolWorkItem.ExecuteWorkItem()
   at bool System.Threading.ThreadPoolWorkQueue.Dispatch()
   at bool System.Threading._ThreadPoolWaitCallback.PerformWaitCallback()
```

**Comment by Frooxius** [comment]
Pushed a high priority fix for this 2020.11.9.372! The dumped recovery file should now be save-able, as long as the cache wasn't cleared there should be no data loss. Thanks for the report!

**Comment by Frooxius** [comment]
If you've saved your world recently and the changes didn't go through, you can recover them as long as you haven't cleared your cache. Check any recently saved worlds to see if they're missing any changes. If they are, do the following:

1) Locate the corresponding log file in the Steam installation for when you last saved the world
2) Search for the first occurence of "Exception in the save process!" following "Saving world: <name of your world>". Make sure to find the last one in the log for particular world.
3) Copy the path following "Dumping the raw save data to:" (including the .lz4bson extension) from the log
4) Switch to Neos and paste (e.g. Ctrl+V). This will load the last version of the world that failed to save.
5) Verify that it's the right version and save a copy of it. You can then use this copy to overwrite the proper version (overwrite worlds by clicking on an world orb while holding the world orb you want to overwrite with)
